src/restricted-endpoints/replication/changes/changes.controller.ts

```diff
@@ -0,0 +1,61 @@
+import { Controller, Get, Param, Query, UseGuards } from '@nestjs/common';
+import { ChangesResponse } from '../bulk-document/couchdb-dtos/changes.dto';
+import { CouchdbService } from '../../../couchdb/couchdb.service';
+import { map, Observable } from 'rxjs';
+import { OnlyAuthenticated } from '../../../auth/only-authenticated.decorator';
+import { CombinedAuthGuard } from '../../../auth/guards/combined-auth/combined-auth.guard';
+import { User } from '../../../auth/user.decorator';
+import { UserInfo } from '../../session/user-auth.dto';
+import {
+  DocumentAbility,
+  PermissionService,
+} from '../../../permissions/permission/permission.service';
+import { DatabaseDocument } from '../bulk-document/couchdb-dtos/bulk-docs.dto';
+import { omit } from 'lodash';
+
+@OnlyAuthenticated()
+@UseGuards(CombinedAuthGuard)
+@Controller()
+export class ChangesController {
+  constructor(
+    private couchdbService: CouchdbService,
+    private permissionService: PermissionService,
+  ) {}
+
+  /**
+   * Get the changes stream.
+   * The changes feed only returns the doc IDs to which the requesting user has access.
+   * Even if `include_docs: true` is set, the stream will not return the document content.
+   * @param db
+   * @param params
+   * @param user
+   */
+  @Get(':db/_changes')
+  changes(
+    @Param('db') db: string,
+    @Query() params,
+    @User() user: UserInfo,
+  ): Observable<ChangesResponse> {
+    return this.couchdbService
+      .get<ChangesResponse>(db, '_changes', {
+        ...params,
+        include_docs: true,
+      })
+      .pipe(map((res) => this.filterChanges(res, user)));
+  }
+
+  private filterChanges(
+    changes: ChangesResponse,
+    user: UserInfo,
+  ): ChangesResponse {
+    const ability = this.permissionService.getAbilityFor(user);
+    changes.results = changes.results
+      .filter((change) => this.canReadDoc(change.doc, ability))
```

Comment:
At the moment the initial `sync` call (not live sync) in the frontend finishes the first time an empty changes feed is returned.
This does not really create a problem as the remaining docs are synced through the live sync later.
We could however further reduce the amount of requests done by the frontend by already requesting and returning further changes here in the backend whenever not enough results remain in the returned changes.

Some questions remain, e.g. whether we always want to remain within the limit or whether we just ensure that at least limit docs are returned (but possibly more) to be more performant.

Reply:
I think it would be good to do this not only for performance but also for cases that expect this to closely follow official specifications, e.g. make sure this also works without live sync.

I guess doing a "simple" recursive call in `changes()` to loop until we reach the end of the overall changes or the limit and cut of the returned result exactly at the limit wouldn't be extremely difficult?

Reply:
Not that simple but I see the point of following the specs a bit closer.